app/uk/gov/hmrc/agentsubscription/controllers/SubscriptionController.scala

```diff
@@ -16,46 +16,61 @@
 
 package uk.gov.hmrc.agentsubscription.controllers
 
-import javax.inject._
 import com.kenshoo.play.metrics.Metrics
+import javax.inject._
 import play.api.libs.json.Json.toJson
-import uk.gov.hmrc.agentsubscription.connectors.{ AuthActions, MicroserviceAuthConnector }
+import play.api.libs.json.{ JsError, JsSuccess }
+import play.api.mvc.{ Action, AnyContent }
+import uk.gov.hmrc.agentsubscription.auth.AuthActions
+import uk.gov.hmrc.agentsubscription.connectors.MicroserviceAuthConnector
 import uk.gov.hmrc.agentsubscription.model.{ SubscriptionRequest, SubscriptionResponse, UpdateSubscriptionRequest }
 import uk.gov.hmrc.agentsubscription.service.{ EnrolmentAlreadyAllocated, SubscriptionService }
+import uk.gov.hmrc.agentsubscription.utils.toFuture
 import uk.gov.hmrc.http.Upstream5xxResponse
 import uk.gov.hmrc.play.microservice.controller.BaseController
-import uk.gov.hmrc.play.http.logging.MdcLoggingExecutionContext.fromLoggingDetails
-import uk.gov.hmrc.play.HeaderCarrierConverter.fromHeadersAndSession
 
 @Singleton
 class SubscriptionController @Inject() (subscriptionService: SubscriptionService)(implicit
   metrics: Metrics,
   microserviceAuthConnector: MicroserviceAuthConnector)
   extends AuthActions(metrics, microserviceAuthConnector) with BaseController {
 
-  def createSubscription = affinityGroupAndEnrolments { implicit request => implicit authIds =>
-    implicit val hc = fromHeadersAndSession(request.headers, None)
-    withJsonBody[SubscriptionRequest] { subscriptionRequest =>
```

Comment:
What's wrong with withJsonBody?

Reply:
earlier AuthActions class was dealing with parsing request json which it should not worry about....  `Action.async(parse.json)` 
its not Auth class responsibility to parse the request, so refactored it .. :)